client/visitor.go

```diff
@@ -328,3 +335,211 @@ func (sv *XtcpVisitor) handleConn(userConn net.Conn) {
 	frpIo.Join(userConn, muxConnRWCloser)
 	xl.Debug("join connections closed")
 }
+
+type SudpVisitor struct {
+	*BaseVisitor
+
+	// udpConn is the listener of udp packet
+	udpConn *net.UDPConn
+
+	// data transfer connection, should be keep-lived，if timeout or break，
+	// it should be reconnect to frps and release old connection
+	workConn net.Conn
+
+	sendCh chan *msg.UdpPacket
+	readCh chan *msg.UdpPacket
+
+	checkCloseCh chan int
+
+	cfg *config.SudpVisitorConf
+}
+
+// SUDP Run start listen a udp port
+func (sv *SudpVisitor) Run() (err error) {
+	xl := xlog.FromContextSafe(sv.ctx)
+
+	addr, err := net.ResolveUDPAddr("udp", fmt.Sprintf("%s:%d", sv.cfg.BindAddr, sv.cfg.BindPort))
+	if err != nil {
+		return
+	}
+	sv.udpConn, err = net.ListenUDP("udp", addr)
+	if err != nil {
+		xl.Warn("listen udp port %s error: %v", addr.String(), err)
+		return
+	}
+
+	go sv.worker()
+	return
+}
+
+func (sv *SudpVisitor) worker() {
+	xl := xlog.FromContextSafe(sv.ctx)
+	xl.Debug("starting sudp proxy")
+
+	sv.sendCh = make(chan *msg.UdpPacket, 1024)
+	sv.readCh = make(chan *msg.UdpPacket, 1024)
+	sv.checkCloseCh = make(chan int)
+
+	// udp service -> frpc -> frps -> frpc visitor -> user
+	workConnReaderFn := func(conn net.Conn) {
+		for {
+			var (
+				rawMsg msg.Message
+				errRet error
+			)
+
+			conn.SetReadDeadline(time.Now().Add(60 * time.Second))
+			if rawMsg, errRet = msg.ReadMsg(conn); errRet != nil {
+				xl.Warn("read from workconn for user udp conn error: %v", errRet)
+				conn.Close()
+
+				errors.PanicToError(func() {
+					sv.checkCloseCh <- 1
```

Comment:
Why not just close this channel?

Reply:
close(checkCloseCh) just be called by Close() function and will close sudp proxy.

sv.checkCloseCh <- 1 just for
```
visitorConn, err := sv.getNewVisitorConn()
```
to continue get visitorConn